### DISCUSSION TITLE
Transmission

### DISCUSSION THREAD

**Discussion by hao14chasing** [ymahlau/fdtdx#238] [Q&A]
How to get transmission data from the detectors? In documentation, it specifically points to `ModeOverlapDetector`, but it doesn't normalized by source's power. I wonder how can we also get source's power so that we can actually calculate the transmission.

**Comment by ymahlau** [comment]
Hi @hao14chasing, you can do that with the ModeOverlapDetector. The sources in fdtdx are normalized to output a poynting flux of 1, so there is no normalization necessary. For example, if you measure a poynting flux of 0.96 with the ModeOverlapDetector, then you have 96% transmission.

**Reply by hao14chasing** [reply]
Hi, thank you very much for your reply.

If `ModeOverlapDetector` is already normalized internally, then I think I may have misunderstood something in my simulation setup. I’ve been debugging this issue for quite a long time without success, and I would really appreciate any guidance if possible.

I’m currently trying to simulate a symmetric spectrometer device in **FDTDX** and cross-validate the results against **Lumerical FDTD**.

---

### FDTDX implementation

Here is my implementation in FDTDX (notebook [link](https://github.com/hao14chasing/Test_FDTDX/blob/main/Spectrometer/FDTDX%20Scripts.ipynb) for reference):

The geometry and setup look like this:

<img width="1412" height="865" alt="image" src="https://github.com/user-attachments/assets/c0c87baf-164b-4dfc-b244-7c0c85092c69" />

<img width="1597" height="490" alt="image" src="https://github.com/user-attachments/assets/ff460928-ae0d-4928-8fca-b4215c68ab09" />

---

### Lumerical implementation

I also implemented the same structure in Lumerical using `lumapi` (script [link](https://github.com/hao14chasing/Test_FDTDX/blob/main/Spectrometer/Lumerical%20Scripts.ipynb) here):


<img width="320" height="511" alt="image" src="https://github.com/user-attachments/assets/02ed318a-b38b-47f7-8b09-10701778b539" />

In the Lumerical GUI, I manually adjusted:

* PML layers = 10 (to match the PML thickness in FDTDX)
* Mode source = fundamental mode only (to align with FDTDX)

---

### Transmission results discrepancy

After running both simulations, I observe a major discrepancy:

#### **Lumerical transmission**

The transmission spectrum is highly symmetric, as expected from the symmetric device geometry (for example, `detector_00` aligns well with `detector_11`.):

<img width="1404" height="965" alt="image" src="https://github.com/user-attachments/assets/c9b2922a-ae22-4b15-93ce-cf756597ca4b" />

#### **FDTDX transmission**

However, in FDTDX I obtain:

<img width="1589" height="861" alt="image" src="https://github.com/user-attachments/assets/ec42e740-43b5-456c-becd-5eed2dd468db" />

This result is:

* **anti-symmetric** rather than symmetric
* about **1–2 orders of magnitude smaller** (≈1e−4) compared with Lumerical (≈1e−2 to 1e−3)

Even when I shift the detector sampling window later (e.g., starting at `0.99 * config.time` for a maximum transimission value), the transmission remains much smaller.

---

### Possible causes I’m considering

1. **Broadband vs single-frequency excitation**

   Although I'm using mode source in both cases, there's a difference in souces definition.

   In Lumerical, I sweep wavelengths from 1.54–1.56 µm with a Gaussian broadband frequency.

   In FDTDX, since broadband excitation does not seem fully supported (to my knowledge), I run one wavelength per simulation using `SingleFrequencyProfile`.

   This difference might contribute to the discrepancy.

2. **Mode normalization / field amplitude mismatch**

   When visualizing the fundamental mode, I notice differences in magnitude:

<img width="1575" height="797" alt="image" src="https://github.com/user-attachments/assets/a9cc73f3-85b9-45f7-b572-6c7fc72f68a0" />

   Both solvers identify a TM-like mode, but the mode amplitude in FDTDX appears much smaller than in Lumerical.

   I’m unsure whether this is expected due to normalization conventions, or whether it could explain the reduced transmission.
   (In the documentation examples, the mode magnitude also looks small but the reported transmission is correct, so I’m confused.)

---

I would greatly appreciate any guidance on whether I may be missing an important normalization step, or if there could be a bug or mismatch in how overlap-based transmission is computed in this setup.

Thank you again for your time and help.

**Reply by ymahlau** [reply]
Hi @hao14chasing, we will take a look at this, though it might take a few days (I have vacation part of next week). @l-berg could you help me take a look at this?

**Reply by l-berg** [reply]
Hi @hao14chasing, thank you for the detailed report. Here are some potential issues I found in your notebook.

1. The mode sources and mode detectors use the slice they are placed over to compute the eigen-modes. However, your mode detectors do not extend beyond the waveguides. Thus, their mode computation results in unexpected modes and they do not measure the modes you are interested in. Enlarging the shape to include background material would fix this issue (e.g., set `partial_grid_shape=(1, 20, 50)` just like the source).

2. I think there is a misunderstanding regarding the OnOffSwitch interface. Currently, the detectors start measuring the mode at 0.99*simulation_time = 99e-15 seconds and stop when the simulation ends at 100e-15 seconds. Since the period length for your 1.5 um light is over 5e-15 seconds, 40 periods would mean over 200e-15 seconds, which of course is not possible when only simulating for 100e-15 seconds. Consider using `end_time` instead, and a shorter on-time or a significantly longer simulation. (We should maybe add some assert statements to catch this early in the future.)
```
fdtdx.ModeOverlapDetector(
    ...
    switch=fdtdx.OnOffSwitch(period=period, start_time=0.99*config.time, on_for_periods=40)
)
```

3. Consider extending the outgoing waveguides all the way to the edge of the simulation into the PML layers.

Hope this helps.

**Reply by hao14chasing** [reply]
Hi @l-berg and @ymahlau, thank you for the suggestion. I have updated my code following your recommendation and am now cross-verifying the results between **FDTDX** and **Lumerical** (before doing optimization) using a simplified setup with **two output channels at a single wavelength**.

So far, both simulators produce **symmetric power distribution between the two output waveguides**. However, I observe a discrepancy between FDTDX and Lumerical when increasing the **output channel separation**, as shown in the figure below:

<img width="752" height="452" alt="image" src="https://github.com/user-attachments/assets/25312413-4086-43b7-8f7a-ea27d760d3bc" />

In Lumerical, when the separation increases (around **0.5 µm**), the transmission first increases and then decreases. My interpretation is that when the waveguides are very close, there may be **mode cross-talk or modal overlap**, which reduces the coupling efficiency. Increasing the separation initially reduces this interaction and improves transmission.

However, in **FDTDX**, the transmission appears to **decrease monotonically** as the separation increases.

For **small separations (~0.1 µm)**, where I performed the topology optimization in FDTDX, the results roughly agree with Lumerical. But at **larger separations**, the discrepancy becomes more noticeable.

I am therefore wondering whether the discrepancy may be due to the fact that FDTDX might not fully account for modal overlap (or coupling) between waveguides.

If you have any suggestions on what might cause this difference, I would greatly appreciate your insight.

**Reply by ymahlau** [reply]
Hi @hao14chasing, It is a bit difficult for me to comment on this without seeing the actual simulation setup. Are you measuring transmission in each waveguide individually in both fdtdx and lumerical and reporting the sum in the image? If so, I don't see why there should be a difference in the handling of mode overlap. 

You are right that when waveguides are very close there may be interactions between the waveguide modes. You could check that by calculating the output mode over both waveguides (using a large detector that covers both waveguides). If the mode looks different than the individual modes you have interactions.

**Reply by hao14chasing** [reply]
Thank you for the reply. Yes, I am measuring each output waveguide individually in both cases.

For reference, the scripts are here:

* FDTDX: [Notebook link](https://github.com/hao14chasing/Test_FDTDX/blob/main/FDTDX%20Scripts.ipynb)
* Lumerical: [Notebook link](https://github.com/hao14chasing/Test_FDTDX/blob/main/Lumerical%20Scripts.ipynb)

With those scripts, FDTDX gives:

<img width="545" height="391" alt="image" src="https://github.com/user-attachments/assets/ae41035d-8650-4d40-bbd7-54e43b04aae0" />

and Lumerical gives:

<img width="590" height="390" alt="image" src="https://github.com/user-attachments/assets/0f8f79e3-5c2b-4919-8767-46edaaa83a15" />

After looking through the setup more carefully, I do not think there is an obvious bug in `fdtdx`, but I am wondering whether the two tools may currently be reporting slightly different observables.

My understanding is that in `fdtdx`, `ModeOverlapDetector` computes a mode on the detector slice itself and then evaluates overlap against that mode. In contrast, in my current Lumerical script I am using a DFT monitor, which is measuring total power flux through the monitor plane rather than an explicit modal overlap. If that is correct, then the comparison may currently be:

* `fdtdx`: how much of the field matches the local guided mode
* `Lumerical`: how much total power passes through the monitor plane

So one possibility is that the Lumerical monitor is also counting some non-guided or radiative power within the monitor window, while `fdtdx` is more strictly projecting onto the guided mode.

Would that interpretation make sense? If so, then perhaps a better apples-to-apples comparison would be either:

1. use a flux-type detector in `fdtdx` and compare flux to flux, or
2. use a mode expansion / port-mode measurement in Lumerical and compare mode overlap to mode overlap.

**Reply by hao14chasing** [reply]
After looking through the setup more carefully, I no longer think the issue is simply that I was comparing a mode-overlap quantity in `fdtdx` against a total-flux quantity in Lumerical. Initially that seemed plausible, since my first Lumerical setup used a DFT monitor, while `fdtdx` uses `ModeOverlapDetector`, which computes overlap against a mode solved on the detector slice itself.

However, after replacing the Lumerical DFT monitor with a **mode expansion monitor**, I still see a different trend in Lumerical, including a slight increase in transmission for intermediate output gaps:

<img width="590" height="390" alt="image" src="https://github.com/user-attachments/assets/9a6573d8-64a8-4ae3-bd17-2e968f447995" />

I checked the `fdtdx` source again and did not find an obvious issue in `ModeOverlapDetector.compute_overlap()` itself. The overlap expression there looks standard. What stood out to me instead is the source implementation: in `ModePlaneSource.apply()`, a complex mode is first computed, but then the injected fields are explicitly reduced to

```python
jnp.real(mode_E), jnp.real(mode_H)
```

before being launched.

Since `ModeOverlapDetector`, on the other hand, later projects the simulated fields onto the full complex mode, I am wondering whether this source/detector inconsistency could affect the measured overlap and possibly also its trend versus output gap.

A second point is that `fdtdx` seems to return a raw overlap coefficient from `compute_overlap()`, namely something of the form

```python
alpha_coeff = jnp.sum(E_cross_H_star_sim + E_star_cross_H_sim)
alpha_coeff = alpha_coeff / 4.0
```

and then the reported quantity is typically taken as

```python
jnp.abs(det.compute_overlap(...))
```

So in `fdtdx`, this appears to be a raw overlap-like amplitude constructed from the phasor fields and the detector mode. By contrast, Lumerical’s mode expansion monitor reports quantities such as `Tforward`, which are already normalized modal transmission results in Lumerical’s convention.

Because of that, I understand that the absolute values from the two tools are not necessarily expected to match exactly unless the normalization conventions are aligned. That part would explain an absolute offset. What I am less sure about is whether the `jnp.real(...)` step in `ModePlaneSource.apply()` could also be contributing to the different *shape* of the transmission-vs-gap curve.

Would that be a reasonable interpretation, or am I missing something else in how `ModeOverlapDetector` is intended to be normalized/used?

**Reply by ymahlau** [reply]
Hi @hao14chasing, sorry for the late reply, I did not have time to look at this into detail. I am also not sure where the differences come from. If I find the time, I will investigate a bit more. In general, the use of jnp.real is fine since fdtdx does not implement dispersion and all fields are real always.

Some ideas may be: 
- Do you record the same mode in fdtdx and lumerical (TE / TM and same order)? 
- There is a small bug in the field interpolation in fdtdx (see #256). I dont think that this should make such a big difference here, but could contribute to differences

If you have any new insights please let me know!

**Comment by ymahlau** [comment]
@renaissancenerd, could you please take a look if you find anything here? For example, how do the results differ between TE and TM mode with the fdtdx script provided? Also please check if you find any mistakes in the script that may contribute to the difference reported here.

**Reply by renaissancenerd** [reply]
@ymahlau I suspect the `ModePlaneSource` to be a root cause because it uses `mode_index=0.` fdtdx defines TE modes as those where the electric field component along the propagation direction is zero — so for x-propagating modes, TE means Ex = 0, meaning the dominant E-field is in the y-direction (in-plane). Without a filter_pol filter, there is no guarantee that `mode_index=0` picks the same polarization that Lumerical's `updatesourcemode(1)` picks. As per our docs, correct method would be to apply `filter_pol="te"` to both the source and all detectors so they all operate on the same polarization and match Lumerical's default

**Reply by hao14chasing** [reply]
Hi @renaissancenerd, thank you for the advice. For my waveguide at 1.55 um, the fundamental mode is `TM`, and both fdtdx and Lumerical give the same mode. I tried adding `filter_pol="tm"`, but results stay the same for fdtdx.

**Comment by hao14chasing** [comment]
Hello, is there any updates on this?

**Reply by ymahlau** [reply]
Hi @hao14chasing, I did not have time to take a closer look at this. I am currently on a work trip and might find the time to take a closer look in the next 2-3 weeks, but cannot promise anything

**Reply by hao14chasing** [reply]
Thanks for the update and for the care you're putting into this. I really appreciate it—there is absolutely no rush on this end, so please take the time you need.
